### DISCUSSION TITLE
Adventure map, AI switches between heroes too fast

### DISCUSSION THREAD

**Discussion by Branikolog** [ihhub/fheroes2#3009] [General]
![image](https://user-images.githubusercontent.com/55348946/112374765-8b7b7e80-8cf3-11eb-8906-3920c45f20d2.png)
AI has multiple heroes. One of them stands close to a gold pile (a hero close to the left edge) while AI turn starts with another hero(In center).
A hero in center captures Alchemist Lab, then AI switches to the left hero, collects gold pile and returns back to the first hero. But this happens sooo fast, so I even couldn't notice anything. The Only thing I've noticed is that after first hero captures Alchemist Lab one frame on a screen changes to something different (looks like one frame of game was corrupted). So that's why I started investigating this issue.
[O_O.zip](https://github.com/ihhub/fheroes2/files/6200166/O_O.zip)
(This happens not always, so try to reload and end turn a few times: you will notice something on a screen, while AI hero captured Alchemist Lab)
This looks not good. I think we should either play animation of gold pile (and any other resource or artifact) being disappeared after AI hero took it. 
Or we can try to force the game to make a little pause while switching between different heroes. (~0.5-1 sec or smth like that) so this would prevent screen from being redrawn for 1/60sec for focusing on another hero and returning back to the first one.

**Comment by LeHerosInconnu** [comment]
Here are two videos to show the problems:

Captured with OBS Studio:

https://user-images.githubusercontent.com/43583994/112532663-61849380-8da9-11eb-97cc-2655734bcd6d.mp4

Video capture of the monitor:
I will delete the file in 30 days (file about 82MB).
https://drive.google.com/file/d/1hVHjaBTiaM8KufCD7GZg9SThIAPiAmKe/view?usp=sharing

The pause between the heroes' animations is not necessarily needed.
We can notice that the animation of the hero's movement starts before the screen is centered on the hero.
That's why we don't see the hero of the green player near the obelisk picking up the Gold pile.

First transition (video capture of the monitor):
The hero has already started his movement.
![Frame-01536](https://user-images.githubusercontent.com/43583994/112534132-02c01980-8dab-11eb-8c8b-dbba5727c1be.png)
![Frame-01537](https://user-images.githubusercontent.com/43583994/112534770-c7721a80-8dab-11eb-8f7f-42617feed811.png)
![Frame-01538](https://user-images.githubusercontent.com/43583994/112535356-80385980-8dac-11eb-9afa-9e0337e7f4ff.png)

Second transition (video capture of the monitor):
The hero has already picked up the pile of Gold.
![Frame-01662](https://user-images.githubusercontent.com/43583994/112536584-f6898b80-8dad-11eb-9eb1-89fe6f7552a0.png)
![Frame-01663](https://user-images.githubusercontent.com/43583994/112537141-a52dcc00-8dae-11eb-8e3a-824e24d94ab9.png)
![Frame-01664](https://user-images.githubusercontent.com/43583994/112537692-4a48a480-8daf-11eb-9fe6-cf27fe9cca10.png)

Hero animation already started (from both videos):
![Frame-00092 mod 01](https://user-images.githubusercontent.com/43583994/112540013-fab7a800-8db1-11eb-9c83-f1a438fd548e.png)
![Frame-00219 mod 01](https://user-images.githubusercontent.com/43583994/112540057-06a36a00-8db2-11eb-965c-aee5ef3ffeeb.png)
![Frame-00314 mod 01](https://user-images.githubusercontent.com/43583994/112540104-13c05900-8db2-11eb-8af1-10e1a3709989.png)
![Frame-00411 mod 01](https://user-images.githubusercontent.com/43583994/112540142-1e7aee00-8db2-11eb-9523-088f1e3ace53.png)
![Frame-00488 mod 01](https://user-images.githubusercontent.com/43583994/112540172-263a9280-8db2-11eb-9fb7-4df0b4ee5c1c.png)
![Frame-00623 mod 01](https://user-images.githubusercontent.com/43583994/112540198-2dfa3700-8db2-11eb-91c7-6922805714ce.png)
![Frame-00673 mod 01](https://user-images.githubusercontent.com/43583994/112540223-35b9db80-8db2-11eb-9e4b-df978bac0c4d.png)
![Frame-01538 mod 01](https://user-images.githubusercontent.com/43583994/112540247-3d798000-8db2-11eb-99c9-23a9faa69384.png)
![Frame-01664 mod 01](https://user-images.githubusercontent.com/43583994/112540271-44a08e00-8db2-11eb-8821-401cefb9b1c5.png)
![Frame-01760 mod 01](https://user-images.githubusercontent.com/43583994/112540292-4b2f0580-8db2-11eb-934f-13ab4f5bb6c7.png)
![Frame-01855 mod 01](https://user-images.githubusercontent.com/43583994/112540316-52561380-8db2-11eb-97af-03ee26406f03.png)
![Frame-01933 mod 01](https://user-images.githubusercontent.com/43583994/112540335-597d2180-8db2-11eb-8fa1-0c2b9d840f0b.png)
![Frame-02068 mod 01](https://user-images.githubusercontent.com/43583994/112540361-5f730280-8db2-11eb-9a29-87092a65f4e2.png)
![Frame-02118 mod 01](https://user-images.githubusercontent.com/43583994/112540388-669a1080-8db2-11eb-9e2c-c3136ba94ff8.png)

Glitch when the hero captures the Alchemist Lab (video capture of the monitor):
![Frame-01550 mod 01](https://user-images.githubusercontent.com/43583994/112540780-e031fe80-8db2-11eb-8cdc-78ce1d220161.png)
![Frame-01551 mod 01](https://user-images.githubusercontent.com/43583994/112540805-e7590c80-8db2-11eb-8664-e0ef54932aa9.png)
![Frame-01552 mod 01](https://user-images.githubusercontent.com/43583994/112540824-ee801a80-8db2-11eb-8ea9-f946481232f0.png)
![Frame-01553 mod 01](https://user-images.githubusercontent.com/43583994/112540847-f50e9200-8db2-11eb-8a3b-00803e65dac2.png)
![Frame-01554 mod 01](https://user-images.githubusercontent.com/43583994/112540870-fb9d0980-8db2-11eb-96e7-d90aee2d5c1e.png)
![Frame-01555 mod 01](https://user-images.githubusercontent.com/43583994/112540890-02c41780-8db3-11eb-82d5-8f90b6042090.png)

The hero seems to go backwards (from OBS Studio):
![Frame-00105 mod 02](https://user-images.githubusercontent.com/43583994/112541465-b4634880-8db3-11eb-8e8c-837671022d65.png)
![Frame-00106 mod 02](https://user-images.githubusercontent.com/43583994/112541490-bcbb8380-8db3-11eb-8e49-0a11ecbc9737.png)
![Frame-00107 mod 002](https://user-images.githubusercontent.com/43583994/112541526-c80eaf00-8db3-11eb-925b-4b70fd7021d2.png)
![Frame-00108 mod 02](https://user-images.githubusercontent.com/43583994/112541568-d230ad80-8db3-11eb-87aa-8395d1523c65.png)
![Frame-00109 mod 02](https://user-images.githubusercontent.com/43583994/112541595-da88e880-8db3-11eb-8413-540c6dfeb04a.png)

Also I suspect that the animation of the previous hero is not yet finished when the animation of the next hero starts.
Related issue: https://github.com/ihhub/fheroes2/issues/2169

**Reply by Branikolog** [reply]
@LeHerosInconnu 
Oh, thanks for the precision!
Do you think, to prevent this glitch we simply need to start AI heroes' movement at the right moment?
What do you think about adding (probably faster than normal) animation of resources being vanished?

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

> @LeHerosInconnu
> Oh, thanks for the precision!
> Do you think, to prevent this glitch we simply need to start AI heroes' movement at the right moment?

It should be done this way:
- The animation of the first hero's movement should be played entirely.
- Possibly a very short pause if necessary.
- The screen view is positioned on the next hero.
- Possibly a very short pause if necessary.
- The animation of the second hero's movement should be played entirely.
- Possibly a very short pause if necessary.
- The screen view is positioned on the next hero.
- Possibly a very short pause if necessary.
And so on.

> What do you think about adding (probably faster than normal) animation of resources being vanished?

This should not cause an excess of waiting time for the human player either.
This is okay in the original game when everything is displayed correctly.

**Reply by Branikolog** [reply]
I think, if we add like ~0.2 sec for every resource it won't drastically change the AI turn's duration. But the game would look a bit more pleasant.

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> I think, if we add like ~0.2 sec for every resource it won't drastically change the AI turn's duration. But the game would look a bit more pleasant.

Not sure.
Let's say 5 AI players with 8 heroes each, who each collect 10 piles of resources, that's 5 x 8 x 10 x 0.2 sec = 80 seconds. :(

**Reply by Branikolog** [reply]
@LeHerosInconnu 
Have you Ever saw AI collect more than 10-15 items during its turn? :D 
Actually, the lacking of this animation for me personally looks a bit weird. When AI hero captures a mine, with a few piles of resources nearby and then collects these resources standing in one place it looks not very well for me. It looks like hero starts spinning in a single place and resources disappear in less than a second.
In H3 during AI turn we can watch that vanishing items on a world map with normal speed. I just want something same... A bit faster, probably. (As in H3 it's like ~0.7sec animation)

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> @LeHerosInconnu
> Have you Ever saw AI collect more than 10-15 items during its turn? :D

All possibilities should be considered. :)

> Actually, the lacking of this animation for me personally looks a bit weird. When AI hero captures a mine, with a few piles of resources nearby and then collects these resources standing in one place it looks not very well for me. It looks like hero starts spinning in a single place and resources disappear in less than a second.
> In H3 during AI turn we can watch that vanishing items on a world map with normal speed. I just want something same... A bit faster, probably. (As in H3 it's like ~0.7sec animation)

In the original game, the AI turn (calculations) is very fast, almost instantaneous, and the hero movement animations are also fast, at most a few seconds and that even at the slowest animation speed.
I wouldn't want to wait more than a few seconds all together during the AI turn.
This works well for me in the original game.
The game should remain very fast, that is part of the fun of playing Heroes 2.

**Reply by Branikolog** [reply]
I'd prefer to watch a little pause for vanishing of resources. Now it seems to disappear in a single frame so I can't even notice what type of resource AI gathered...

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> I'd prefer to watch a little pause for vanishing of resources. Now it seems to disappear in a single frame so I can't even notice what type of resource AI gathered...

There are already slower animation speeds for AI heroes that allow the player to see what the opposing heroes are doing more quietly.
Isn't it enough?

**Reply by Branikolog** [reply]
At lower speed it keeps spinning all the same. )

**Reply by LeHerosInconnu** [reply]
@Branikolog.

> At lower speed it keeps spinning all the same. )

Maybe this could be added but only for the slowest animation speed for AI heroes?
(Or as an option, as a last resort.)

**Comment by LeHerosInconnu** [comment]
I just observed the same phenomenon with version 1.0.1 6678.

The first purple hero captures the alchemist lab, the other purple hero to the south captures the treasure chest, the first purple hero moves away from the alchemist lab, but it happens so quickly that it looks like there is a graphic glitch (the treasure chest capture is not visible).

A delay should be added when the game switches from one hero to another (just after the game moves to the next hero and therefore before the hero's action starts).

In fheroes2:

![vlcsnap-2023-02-25-20h08m13s087](https://user-images.githubusercontent.com/43583994/221376123-8cec17ad-3515-494d-a168-84cc22c31fd4.png)

![vlcsnap-2023-02-25-20h08m26s014](https://user-images.githubusercontent.com/43583994/221376126-c0624e64-b5b8-4937-91b8-15cc11ee244d.png)

https://user-images.githubusercontent.com/43583994/221376100-eee9d56d-d551-4def-80a6-0a3bef80519f.mp4

Here is the save file:
[Graphic glitch ai turn 01.zip](https://github.com/ihhub/fheroes2/files/10832114/Graphic.glitch.ai.turn.01.zip)

Another example, where we don't have time to see that the teleporting purple hero captures an artifact and a bunch of resources next to the wandering group of titans.

In fheroes2:

https://user-images.githubusercontent.com/43583994/221376213-6fc96e10-8049-4718-8ef6-0227b9c85f51.mp4

Here is the save file:
[Quick ai hero 01.zip](https://github.com/ihhub/fheroes2/files/10832118/Quick.ai.hero.01.zip)

**Comment by LeHerosInconnu** [comment]
Another example.
In this one, Ergon attacks Blackburn, and the screen is not positioned on the scene of the action before the defeat window appears.

In fheroes2:

https://user-images.githubusercontent.com/43583994/221547953-d8e56338-caf3-4138-b9f9-40fb81064536.mp4

Here is the save file (just end the turn):
[Quick AI attack castle 01.zip](https://github.com/ihhub/fheroes2/files/10838970/Quick.AI.attack.castle.01.zip)
